### PR TITLE
External Plugin Service (grpc)

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
[RFC](https://hackmd.io/k_hMtUsGTbKl2IksC3IjkA)
[Dockerfile](https://gist.github.com/pingsutw/66b2de83b0114911f85182bf88b10787) for out of core plugin.
[Deployment](https://gist.github.com/pingsutw/0bbfca3600818f0742f78d25cfc22f69) yaml for out of core plugin.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3282

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by honnix** [comment]
This looks great. If I understand it correctly, there needs to be some change on propeller side to initiate RPCs. Is there an open PR for that?

**Comment by pingsutw** [comment]
@honnix [here](https://github.com/flyteorg/flyteplugins/pull/330). I'm adding backend plugin service to the helm chart and single binary (demo cluster). are you interested in adding a new plugin in the flytekit? currently, we only have BQ plugin. your feedbacks can help us improve current interface.

**Comment by honnix** [comment]
> @honnix [here](https://github.com/flyteorg/flyteplugins/pull/330). I'm adding backend plugin service to the helm chart and single binary (demo cluster). are you interested in adding a new plugin in the flytekit? currently, we only have BQ plugin. your feedbacks can help us improve current interface.

@pingsutw Thank you for the context. I unfortunately don't have a proper candidate plugin to work on, so I will share some quick reflections.

As I understand, the approach is to have a single gRPC server hosting all plugins with a simple routing based on type. While this keeps the same security model as in-core plugin system when coming to credentials associated with all plugins (thinking about plugins making requests to AWS/GCP), it could be a good opportunity to implement something that can enable hosting plugins in different gRPC servers, so it is both separation of concern and minimizing dependency conflict across plugins. I think that could be achieved by having the capability of configuring different gRPC endpoints with another layer of routing (e.g. a union of types), in the backend plugin implementation. What do you think of this?

**Comment by pingsutw** [comment]
Thanks for the feedback. we were talking about supporting multi grpc endpoint. propeller will have a config allow you to configure different grpc endpoint for different plugin. Therefore, by default, there is a default flyteplugin backend system, and some of the plugins will be created by it. you could also create another deployment that only registers internal plugin (for example, plugins written by java). and you could scale up different deployment independently.

**Comment by honnix** [comment]
Thank you for explaining the plan for propeller and that sounds great 👍

**Comment by kumare3** [comment]
@honnix great question, but just to make things clear,
The idea is that most open source plugins will ship in one plugin container, as they are all blessed and deployment model us easy. Less complexity for users.
But, you are absolutely right, as @pingsutw said, you can always add new endpoints and start a plugin at that endpoint.
This also will help in developing new plugins

**Comment by honnix** [comment]
@kumare3 Yeah it definitely makes sense to ship a collection of managed plugins behind one endpoint to simplify user deployment. We will likely do the same to host our internal managed plugins. Having multiple endpoints makes user developed and hosted plugins achievable, which is great. I think we are very well aligned. Thank you all for explaining and developing this much appreciated feature! We are looking forward to that.

File: plugins/flytekit-bigquery/flytekitplugins/bigquery/backend_plugin.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        if inputs:
+            ctx = FlyteContextManager.current_context()
+            python_interface_inputs = {
+                name: TypeEngine.guess_python_type(lt.type) for name, lt in task_template.interface.inputs.items()
+            }
+            native_inputs = TypeEngine.literal_map_to_kwargs(ctx, inputs, python_interface_inputs)
+
+            job_config = bigquery.QueryJobConfig(
+                query_parameters=[
+                    bigquery.ScalarQueryParameter(name, pythonTypeToBigQueryType[python_interface_inputs[name]], val)
```
Comment (eapolinario) [inline plugins/flytekit-bigquery/flytekitplugins/bigquery/backend_plugin.py:47]: worth validating the input types against https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types ?

**Review by eapolinario — COMMENTED** [review]
Looking good, just a few minor comments.

File: plugins/flytekit-bigquery/flytekitplugins/bigquery/backend_plugin.py (hunk 1 above)
Comment (pingsutw) [reply plugins/flytekit-bigquery/flytekitplugins/bigquery/backend_plugin.py:54]: Thanks. updated it, but not added all the bq types here since not every type can be mapped to python type. btw, bq plugin in propeller only support 4 types. https://github.com/flyteorg/flyteplugins/blob/master/go/tasks/plugins/webapi/bigquery/query_job.go#L218-L255

File: flytekit/clis/sdk_in_container/serve.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    type=int,
+    help="It will wait for the specified number of seconds before shutting down grpc server. It should only be used for testing.",
+)
+@click.pass_context
+def serve(_: click.Context, port, timeout):
+    """
+    Start a grpc server for the external plugin service.
+    """
+    click.secho("Starting the external plugin service...", fg="blue")
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/clis/sdk_in_container/serve.py]: added it, thanks

File: flytekit/models/task.py [hunk 3]
```diff
@@ -329,7 +329,7 @@ def __init__(
         type,
         metadata,
         interface,
-        custom,
+        custom=None,
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/models/task.py]: revert it

File: tests/flytekit/unit/extend/test_backend_plugin.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+dummy_template = TaskTemplate(
+    id=task_id,
+    metadata=task_metadata,
+    interface=interfaces,
+    type="dummy",
+)
+
+
+def test_base_plugin():
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply tests/flytekit/unit/extend/test_backend_plugin.py]: done

File: flytekit/extend/backend/base_plugin.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    State,
+    TaskCreateResponse,
+    TaskDeleteResponse,
+    TaskGetResponse,
+)
+
+from flytekit.models.literals import LiteralMap
+
+
+class BackendPluginBase:
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/extend/backend/base_plugin.py]: done

File: flytekit/extend/backend/base_plugin.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+def convert_to_flyte_state(state: str) -> State:
+    state = state.lower()
+    if state in ["failed"]:
+        return RETRYABLE_FAILURE
+    elif state in ["done", "succeeded"]:
+        return SUCCEEDED
+    elif state in ["running"]:
+        return RUNNING
+    raise ValueError(f"Unrecognize state: {state}")
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/extend/backend/base_plugin.py]: no, this function can be used by any other plugins.

**Comment by pingsutw** [comment]
Add a dockerfile to build an external plugin service, and it will be use in the helm chart.

File: flytekit/extend/backend/base_plugin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (eapolinario) [reply flytekit/extend/backend/base_plugin.py]: As it's currently being used this only works for the states returned by the [bigquery client](https://github.com/flyteorg/flytekit/pull/1524/files#diff-7ff0c3bb68feecc48f1436d3853b9bcf195fa9b06be2fdc044cb3e10d766e9bfR68), no? What happens if other clients have different values for `state` (other than "failed", "done", "succeeded", etc)?

File: .github/workflows/pythonpublish.yml [hunk 7]
```diff
@@ -142,6 +142,48 @@ jobs:
...
+        with:
+          registry: ghcr.io
+          username: "${{ secrets.FLYTE_BOT_USERNAME }}"
+          password: "${{ secrets.FLYTE_BOT_PAT }}"
+      - name: Prepare External Plugin Service Image Names
+        id: external-plugin-service-names
+        uses: docker/metadata-action@v3
+        with:
+          images: |
+            ghcr.io/${{ github.repository_owner }}/flytekit
```
Comment (wild-endeavor) [inline .github/workflows/pythonpublish.yml:169]: let's make a new package for this.  don't want to confuse people with the existing flytekit base image right?

File: flytekit/extend/backend/external_plugin_service.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    def CreateTask(self, request: TaskCreateRequest, context: grpc.ServicerContext) -> TaskCreateResponse:
+        try:
+            req = model.TaskCreateRequest.from_flyte_idl(request)
+            plugin = BackendPluginRegistry.get_plugin(req.template.type)
+            return plugin.create(
+                context=context, inputs=req.inputs, output_prefix=req.output_prefix, task_template=req.template
+            )
+        except Exception as e:
+            context.set_code(grpc.StatusCode.INTERNAL)
+            context.set_details(f"failed to create task with error {e}")
```
Comment (wild-endeavor) [inline flytekit/extend/backend/external_plugin_service.py:26]: should we add logging here too?

File: plugins/flytekit-bigquery/flytekitplugins/bigquery/backend_plugin.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        job_config = None
+        if inputs:
+            ctx = FlyteContextManager.current_context()
+            python_interface_inputs = {
+                name: TypeEngine.guess_python_type(lt.type) for name, lt in task_template.interface.inputs.items()
+            }
+            native_inputs = TypeEngine.literal_map_to_kwargs(ctx, inputs, python_interface_inputs)
+
+            job_config = bigquery.QueryJobConfig(
+                query_parameters=[
```
Comment (wild-endeavor) [inline plugins/flytekit-bigquery/flytekitplugins/bigquery/backend_plugin.py:53]: can we log this?

File: flytekit/extend/backend/base_plugin.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    State,
+    TaskCreateResponse,
+    TaskDeleteResponse,
+    TaskGetResponse,
+)
+
+from flytekit.models.literals import LiteralMap
+
+
+class BackendPluginBase(ABC):
```
Comment (wild-endeavor) [inline flytekit/extend/backend/base_plugin.py:19]: If this is the main Base class that people will be implementing, can we add a lot more comments here?

**Review by wild-endeavor — APPROVED** [review]
I think I just need to read through the RFC, but can you describe somewhere how traffic is sent?  like can you detail out how does the server receives traffic?  To be concrete, can we write out:
* if I run `pyflyte serve` in one terminal window, if i run in another window a local workflow execution that has that task type, does it now get redirected to my pyflyte serve?
* If I run `flytectl demo start` how does that propeller send traffic to a `pyflyte serve` I have running locally?

+1 otherwise, just a bunch of nit comments, except for the docker container.  I think we should use a different docker container name.  I can help get this set up tomorrow with you @pingsutw

File: flytekit/extend/backend/base_plugin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/extend/backend/base_plugin.py]: sorry I mean they have to add new state to this function, and reuse it.
```python
if state in ["failed", "failure"]:
        return RETRYABLE_FAILURE
...
```

File: Dockerfile.external-plugin-service [hunk 11]
```diff
@@ -0,0 +1,10 @@
+FROM python:3.9-slim-buster
+
+MAINTAINER Flyte Team <users@flyte.org>
+LABEL org.opencontainers.image.source=https://github.com/flyteorg/flytekit
+
+ARG VERSION
+RUN pip install -U flytekit==$VERSION \
+	flytekitplugins-bigquery==$VERSION \
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply Dockerfile.external-plugin-service:8]: because I add a new backend plugin for BQ in this pr

File: flytekit/clis/sdk_in_container/serve.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+from flytekit.extend.backend.external_plugin_service import BackendPluginServer
+
+_serve_help = """Start a grpc server for the external plugin service."""
+
+
+@click.command("serve", help=_serve_help)
+@click.option(
+    "--port",
+    default="80",
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/clis/sdk_in_container/serve.py:15]: admin also use 80, maybe set nodeport to 30090
```bash
flyteadmin                   ClusterIP   10.96.119.241   <none>        80/TCP,81/TCP,87/TCP,10254/TCP             2d
```

File: Dockerfile.external-plugin-service [hunk 13]
```diff
@@ -0,0 +1,10 @@
+FROM python:3.9-slim-buster
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply Dockerfile.external-plugin-service:1]: I think we can use a default version first because it's experimental feature. we can add more version when we need it

File: .github/workflows/pythonpublish.yml (hunk 7 above)
Comment (pingsutw) [reply .github/workflows/pythonpublish.yml:169]: renamed it. `flytekit-external-plugin-service:latest` -> `external-plugin-service:latest`

**Comment by pingsutw** [comment]
> if I run pyflyte serve in one terminal window, if i run in another window a local workflow execution that has that task type, does it now get redirected to my pyflyte serve?

Not for now. we should add a new env or introduce a new execution mode, so flytekit knows when to send the request to external plugin service.

> If I run flytectl demo start how does that propeller send traffic to a pyflyte serve I have running locally?

update the default endpoint to `host.docker.internal:80`

File: flytekit/extend/backend/model.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
+from typing import Optional
+
+from flyteidl.service import external_plugin_service_pb2
+
+from flytekit.models import common, task
+from flytekit.models.literals import LiteralMap
+
+
+class TaskCreateRequest(common.FlyteIdlEntity):
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/extend/backend/model.py]: remove it

File: flytekit/extend/backend/base_plugin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (eapolinario) [reply flytekit/extend/backend/base_plugin.py]: what if a client doesn't use those values? In the code now we translate the values returned by the biquery client [directly](https://github.com/flyteorg/flytekit/pull/1524/files#diff-7ff0c3bb68feecc48f1436d3853b9bcf195fa9b06be2fdc044cb3e10d766e9bfR67-R69), right? Imagine the case of a different plugin where the state doesn't match any of the ones in the BQ client.

File: flytekit/clis/sdk_in_container/serve.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (eapolinario) [reply flytekit/clis/sdk_in_container/serve.py:15]: this is the pod port, right? Having this as port 80 might be annoying in local tests.

Also, this will be exposed through a Service, which we're doing in https://github.com/flyteorg/flyte/pull/3454. Reading that PR though I'm don't get why we need to use NodePort services. Isn't the external plugin service accessed only by propeller?

[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/clis/sdk_in_container/serve.py]: with nodeport, people are able to submit the job from local. only for testing.

**Review by wild-endeavor — APPROVED** [review]
